rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,326 @@
+
+We likely want to make Nix use `--filter=tree:0` by default when fetching repositories.
+`--depth 1` is faster, but it doesn't preserve the commit history,
+and it's not clear how it it interacts with GPG signing.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Committers will need to adapt to the new workflow before March of 2022.
+The command they will run is explained in [signing], but further
+information about how to sign commits is out of scope for this RFC.
+
+Relevant documentation for setting up GPG:
+- https://guix.gnu.org/manual/en/html_node/Commit-Access.html
+- https://www.kernel.org/doc/html/latest/process/maintainer-pgp-guide.html
+
+Flake users will need to become aware of the basics of this system
+when using forks of Nixpkgs, as updating such flake inputs
+will be different from before.
+In addition, flake users should also begin taking use of the features
+introduced in this RFC, as they are usable and useful for all flakes.
+
+## Example flake
+
+```nix
+# flake.nix
+{
+  inputs.someflake = {
+    url = "https://github.com/some/flake.git";
+    startsAt = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
+    startsAtKey = "XXXX XXXX XXXX XXXX XXXX  XXXX XXXX XXXX XXXX XXXX";
+    type = "git";
+  };
+
+  outputs = { someflake }: {};
+}
+```
+
+```nix
+# some/flake/authorized-committers.nix
+{
+  committers.some = "XXXX XXXX XXXX XXXX XXXX  XXXX XXXX XXXX XXXX XXXX";
+}
+```
+
+Now if `https://github.com/some/flake.git` receives a new commit that can
+*not* be trusted according to the algorithm specified in [updating],
+we'd get an error like the following if we run `nix flake update`:
+```
+* **Failed trust verification!** 'someflake': 'git+https://github.com/some/flake.git?ref=master&rev=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' -!> 'git+https://github.com/some/flake.git?ref=master&rev=bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb' (Not signed by a trusted party!)
+```
+
+Notably, there is nothing that tells the user how to bypass the check
+using the `insecureUpdate` attribute,
+since if there were, users might very often ignore the error even if valid.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Merging PRs can not be done through GitHub.
+- Setting up GPG signatures will be a bit extra work for the committers.
+- Committers will need to maintain the list of authorized committers.
+- Fetching Nixpkgs with Nix will take around [4 times](#speeding-up-fetching-git) the time, depending on your hardware.
+- Currently building a fork of Nixpkgs (for e.g. a PR) will still work fine, since
+  the latest commit will be trusted when adding a new flake input, or referring
+  to one from the commandline, but updating the flake input will fail
+  since the commits are likely not signed by authorized committers.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Mandating signing only tags is an alternative, though not very useful for most users.
+
+# Unresolved questions
```

Comment:
Stable Nix's channels are not considered at all. These are updated by downloading a tarball snapshot from channels.nixos.org / Amazon S3 and wouldn't benefit from signed commits at all as it stands.

Reply:
Stable Nix channels are to be deprecated. It doesn't make sense to take them into account, when it's not straight-forward, would take a lot of extra time, when they're going to be replaced by flake inputs in maybe a year's time.

Reply:
It would be simpler to replace them with flake inputs rather than backporting this RFC.

Reply:
See https://github.com/NixOS/nix/issues/748#issuecomment-559521872

Reply:
Flakes will take quite a while to stabilize still (i. e. there are two more nix major releases before that happens at the earliest). I don't think it's an acceptable outcome for a RFC to say “you can have a verified nixpkgs, but you'll have to use experimental features for this”. I'd go even further and say it's precisely the point of having experimental features that no critical infrastructure like this is built upon it.

Additionally I see the following problems:

* The success of this RFC depends on flakes succeeding which isn't guaranteed
* Accepting this RFC before flakes are stable / a revised flakes RFC is accepted doesn't make sense. Current flakes development would be negatively hemmed in by normative stipulations made in this PR.

Reply:
I'm not sure I understand.

It seems to me that flakes succeeding *is* guaranteed, since there is no alternative solution to flakes.
The exact format of flakes might still change, but no matter how it changes, this RFC only depends on the fact that flakes can have inputs that are Git repositories, and that inputs can have attributes.

Reply:
> It seems to me that flakes succeeding is guaranteed, since there is no alternative solution to flakes.

And that is the entire problem with the flakes adoption. Users think that it is the best they can get while we did never talk about how to do it differently or what is actually needed. They are likely to succeed but given that they are almost taken for granted is a bad sign. Perhaps we should also skip this RFC and you just start pushing the code out?

Reply:
> It seems to me that flakes succeeding is guaranteed, since there is no alternative solution to flakes.

And of course there _is_ an alternative solution to flakes, for example `niv`, but NIH strikes as usual.

Reply:
`niv` doesn't have feature-parity with flakes AFAICT, though as I explained further below, anything that makes use of `builtins.fetchGit` could make use of this system, however, how `niv` interacts with it isn't within the scope of the RFC, as it's a separate project. Flakes are part of Nix, and specifying how the interaction works is crucial.

As for channels, they would not work with this system without major changes. You could get all users to use `niv` to fetch Nixpkgs, however, but I don't have any concrete idea for how you'd do that.